server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java

```diff
@@ -0,0 +1,228 @@
+    }
+
+    public Map<String, Mapper> getChildren() {
+        return Collections.unmodifiableMap(this.mappers);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(simpleName());
+        builder.field("type", CONTENT_TYPE);
+        if (includeInParent.explicit()) {
+            builder.field("include_in_parent", includeInParent.value());
+        }
+        if (includeInRoot.explicit()) {
+            builder.field("include_in_root", includeInRoot.value());
+        }
+        if (dynamic != null) {
+            builder.field("dynamic", dynamic.name().toLowerCase(Locale.ROOT));
+        }
+        if (isEnabled() != Defaults.ENABLED) {
+            builder.field("enabled", enabled.value());
+        }
+        serializeMappers(builder, params);
+        return builder.endObject();
+    }
+
+    @Override
+    public ObjectMapper merge(Mapper mergeWith, MapperService.MergeReason reason) {
+        if ((mergeWith instanceof NestedObjectMapper) == false) {
+            throw new IllegalArgumentException("can't merge a non nested mapping [" + mergeWith.name() + "] with a nested mapping");
+        }
+        NestedObjectMapper mergeWithObject = (NestedObjectMapper) mergeWith;
+        NestedObjectMapper.Builder builder = new NestedObjectMapper.Builder(simpleName(), indexCreatedVersion);
+        if (reason == MapperService.MergeReason.INDEX_TEMPLATE) {
+            if (mergeWithObject.includeInParent.explicit()) {
+                builder.includeInParent(mergeWithObject.includeInParent);
+            }
+            if (mergeWithObject.includeInRoot.explicit()) {
+                builder.includeInRoot(mergeWithObject.includeInRoot);
+            }
+        } else {
+            if (includeInParent.value() != mergeWithObject.includeInParent.value()) {
+                throw new MapperException("the [include_in_parent] parameter can't be updated on a nested object mapping");
+            }
+            builder.includeInParent(includeInParent);
+            if (includeInRoot.value() != mergeWithObject.includeInRoot.value()) {
+                throw new MapperException("the [include_in_root] parameter can't be updated on a nested object mapping");
+            }
+            builder.includeInRoot(includeInRoot);
+        }
+        for (Mapper child : mappers.values()) {
+            builder.add(new Mapper.Builder(child.name()) {
+                @Override
+                public Mapper build(ContentPath contentPath) {
+                    return child;
+                }
+            });
+        }
+        ContentPath contentPath = new ContentPath();
+        int lastDot = this.name().lastIndexOf(".");
+        if (lastDot != -1) {
+            contentPath.add(this.name().substring(0, lastDot));
+        }
+        NestedObjectMapper toMerge = builder.build(contentPath);
+        toMerge.doMerge(mergeWithObject, reason);
+        return toMerge;
```

Comment:
It's slightly altered because I've changed things so that we use a builder rather than a clone, but the actual content of everything is the same.  There are a bunch more tests for it now as well.

Reply:
alright!